File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        inorder = []
        
        def traverse(node):
            if node is None:
                return
            
            traverse(node.left)
            inorder.append(node.val)
            traverse(node.right)
        
        traverse(root)
        N = len(inorder)
        
        def construct(left, right):
            if left > right:
                return None
                
            if left == right:
                return TreeNode(inorder[left])
        
            middle = (left + right) // 2
            
            current = TreeNode(inorder[middle])
            current.left = construct(left, middle-1)
            current.right = construct(middle+1, right)
            
            return current
        
        return construct(0, N-1)

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* solve2(vector<int>&v, int start, int end){
        if(start>end) return NULL;
        int middle=(start+end)/2;
        TreeNode* root=new TreeNode(v[middle]);
        root->left=solve2(v,start,middle-1);
        root->right=solve2(v,middle+1,end);
        return root;
    }
    
    void solve1(TreeNode* root, vector<int>&v){
        if(!root) return;
        solve1(root->left,v);
        v.push_back(root->val);
        solve1(root->right,v);
    }
    
    TreeNode* balanceBST(TreeNode* root) {
        vector<int>v;
        solve1(root,v);
        return solve2(v,0,v.size()-1);
    }
};